From 943a7c39add4be1456c26448f839cbd1d66ea787 Mon Sep 17 00:00:00 2001
From: Peter Andreas Entschev <peter@entschev.com>
Date: Wed, 30 Oct 2019 06:25:39 -0700
Subject: [PATCH] UCM/CUDA: Skip ucm_cuda_set_ptr_attr when pointer is null

---
 src/ucm/cuda/cudamem.c | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/src/ucm/cuda/cudamem.c b/src/ucm/cuda/cudamem.c
index 078ded45b1..907f7e51db 100644
--- a/src/ucm/cuda/cudamem.c
+++ b/src/ucm/cuda/cudamem.c
@@ -66,6 +66,11 @@ UCM_OVERRIDE_FUNC(cudaHostUnregister,        cudaError_t)
 
 static void ucm_cuda_set_ptr_attr(CUdeviceptr dptr)
 {
+    if ((void*)dptr == NULL) {
+        ucm_trace("skipping cuPointerSetAttribute for null pointer");
+        return;
+    }
+
     unsigned int value = 1;
     CUresult ret;
     const char *cu_err_str;
@@ -161,10 +166,9 @@ CUresult ucm_cuMemAlloc(CUdeviceptr *dptr, size_t size)
     if (ret == CUDA_SUCCESS) {
         ucm_trace("ucm_cuMemAlloc(dptr=%p size:%lu)",(void *)*dptr, size);
         ucm_dispatch_mem_type_alloc((void *)*dptr, size, UCS_MEMORY_TYPE_CUDA);
+        ucm_cuda_set_ptr_attr(*dptr);
     }
 
-    ucm_cuda_set_ptr_attr(*dptr);
-
     ucm_event_leave();
     return ret;
 }
@@ -201,10 +205,9 @@ CUresult ucm_cuMemAllocPitch(CUdeviceptr *dptr, size_t *pPitch,
                   (WidthInBytes * Height));
         ucm_dispatch_mem_type_alloc((void *)*dptr, WidthInBytes * Height,
                                     UCS_MEMORY_TYPE_CUDA);
+        ucm_cuda_set_ptr_attr(*dptr);
     }
 
-    ucm_cuda_set_ptr_attr(*dptr);
-
     ucm_event_leave();
     return ret;
 }
@@ -281,10 +284,9 @@ cudaError_t ucm_cudaMalloc(void **devPtr, size_t size)
     if (ret == cudaSuccess) {
         ucm_trace("ucm_cudaMalloc(devPtr=%p size:%lu)", *devPtr, size);
         ucm_dispatch_mem_type_alloc(*devPtr, size, UCS_MEMORY_TYPE_CUDA);
+        ucm_cuda_set_ptr_attr((CUdeviceptr) *devPtr);
     }
 
-    ucm_cuda_set_ptr_attr((CUdeviceptr) *devPtr);
-
     ucm_event_leave();
 
     return ret;
@@ -319,10 +321,9 @@ cudaError_t ucm_cudaMallocPitch(void **devPtr, size_t *pitch,
     if (ret == cudaSuccess) {
         ucm_trace("ucm_cudaMallocPitch(devPtr=%p size:%lu)",*devPtr, (width * height));
         ucm_dispatch_mem_type_alloc(*devPtr, (width * height), UCS_MEMORY_TYPE_CUDA);
+        ucm_cuda_set_ptr_attr((CUdeviceptr) *devPtr);
     }
 
-    ucm_cuda_set_ptr_attr((CUdeviceptr) *devPtr);
-
     ucm_event_leave();
     return ret;
 }
